### PR TITLE
Add + support

### PR DESCRIPTION
**Pull request by lopezm94**
This Adds support for the '+' operator which was missing. If I add BLAS optimizations when possible, the file will grow and be less comprehensible so it might be good to have a single file per related set of operators. Also, should there be a common file for functions like copy! and cat_tuple?

I ask all of this because I sincerely don't know how to organize packages.

### CODE REVIEW

**Comment by lopezm94** [comment]
@simonbyrne

**Comment by simonbyrne** [comment]
I'm on holiday at the moment, will look at this when I get back.

**Comment by simonbyrne** [comment]
Sorry for the delay in looking at this, it's been a bit hectic. Thanks for doing this, just a few points:

1. `+` acts like `.+`, except that the arguments have to be arrays of the same size, or numbers. So the easiest way to is to probably have something like
```julia
function add!(O::AbstractArray, As::Union{AbstractArray,Number}...)
    s = size(O)
    for A in As
        isa(A, Number) || s == size(A) || throw(DimensionMismatch("dimensions must match"))
    end
    badd!(O, As...)
end
```
(not sure if that works...)

2. It would be great if you could add some tests.

3. This package is probably getting too big for one file, and should probably be split into smaller files. Just copy the parts into the file, then `include` the file from the main one.

4. `copy!` is actually already define in Base (though with the bit above, shouldn't be needed here).

5. As you can tell, I haven't had much time to pay attention to this package. If you're interested in contributing more generally, I would appreciate any help: I'm happy to provide guidance and feedback. That said, there are quite a few changes in the upcoming 0.5 release that could improve how this package could work, the most relevant being:
 * functions now act like types, so you can dispatch on them, avoiding the use of the `Operator` type. 
 * the general `.` suffix broadcasting.

**Comment by lopezm94** [comment]
Is there a priority between package size and speed? This + implementation is faster than the current .+ implementation, so It must be faster than what you are proposing. This new 0.5 features seem interesting is there a place to check 0.5 notes before the actual release didn't understand very well this part the general `. suffix broadcasting`?

I'll start making the changes :+1:

**Comment by lopezm94** [comment]
ok I found the [notes](https://github.com/JuliaLang/julia/blob/master/NEWS.md), there is another thing relevant to the package in there too:

> The assignment operations .+=, .*= and so on now generate calls to broadcast! on the left-hand side (or call to view(a, ...) on the left-hand side if the latter is an indexing expression, e.g. a[...]). This means that they will fail if the left-hand side is immutable (or does not support view), and will otherwise change the left-hand side in-place (#17510, #17546)

**Comment by simonbyrne** [comment]
> Is there a priority between package size and speed? This + implementation is faster than the current .+ implementation, so It must be even faster than what you are proposing.

Ah, I didn't know that. Ideally we want the fastest thing, but I would be keen to know why `.+` is slower.

**Comment by simonbyrne** [comment]
Oh, I misunderstood your code. This looks reasonable: it would be great if you could add some tests though.

**Comment by lopezm94** [comment]
I just added substraction operator and blas optimizations for copy, scaling, addition and substraction. Finally tests for everything new.

**Comment by lopezm94** [comment]
Something worth to notice is `@into! A = B + C` will always be slower than `A = B + C` (when there are few arguments which is almost always), maybe in the future there could be a `@fast! macro` that doesn't necessarily abides to always overwriting.

File: src/InplaceOps.jl [hunk 1]
```diff
@@ -1,17 +1,10 @@
...
-
-immutable Inplace{N} end
-
-immutable Transpose{T}
-    obj::T
-end
-immutable CTranspose{T}
-    obj::T
-end
+isdefined(:__precompile__) && __precompile__(true)
```
Comment (simonbyrne) [inline src/InplaceOps.jl]: This should go before the `module` statement, see http://docs.julialang.org/en/release-0.4/manual/modules/

Comment (lopezm94) [inline src/InplaceOps.jl]: I see, thanks! So [this](https://github.com/JuliaLang/IterativeSolvers.jl/blob/master/src/IterativeSolvers.jl#L3) would be wrong too, right? its from IterativeSolvers.jl

**Comment by lopezm94** [comment]
Is this ready to merge?

File: src/InplaceOps.jl [hunk 2]
```diff
@@ -85,18 +78,22 @@ bsub!{N}(::Type{Inplace{N}}, As...) = broadcast!(-,As[N],As...)
...
 brdiv!(O::AbstractArray, As...) = broadcast!(/,O,As...)
+add!(O::AbstractArray, As...) = _add!(O, As...)
+sub!(O::AbstractArray, As...) = _sub!(O, As...)
 
 replace_t(ex) = esc(ex)
 function replace_t(ex::Expr)
     if ex.head == symbol("'")
         :(op_ctranspose($(esc(ex.args[1]))))
-    elseif ex.head == :(.')
+    elseif ex.head == symbol(".'")
```
Comment (simonbyrne) [inline src/InplaceOps.jl]: This should be `Symbol` now.

File: src/InplaceOps.jl [hunk 3]
```diff
@@ -120,5 +117,4 @@ macro into!(ex)
     Expr(:call,inplace_sym(ex.args[1]),esc(out),[replace_t(a) for a in ex.args[2:end]]...)
 end
 
-
-end # module
+end
```
Comment (simonbyrne) [inline src/InplaceOps.jl]: Any reason to remove this?

**Comment by simonbyrne** [comment]
Thanks, 2 minor changes and then we can merge it. Sorry for the delay.

File: src/InplaceOps.jl (hunk 3 above)
Comment (lopezm94) [inline src/InplaceOps.jl]: hahahaha srry, I do it subconsciously

**Comment by simonbyrne** [comment]
Thanks!

**Pull request merged**